### PR TITLE
Fix sinpi and cospi with Irrational arguments

### PR DESCRIPTION
**Pull request by giordano**
I replaced `typeof(float(x))` with `float(T)`, because after PR #21219 the latter doesn't allocate for `BigInt` and `Rational`.

### CODE REVIEW

**Comment by giordano** [comment]
Uhm, ok, maybe this fix wasn't really good.

**Comment by giordano** [comment]
PR updated as suggested above.  Now the tests should pass.

File: base/irrationals.jl
Comment (Sacha0) [inline base/irrationals.jl]: Could this signature simplify to `sinc(z::Complex{<:Irrational})`, and likewise for the similar `cosc` definition below?

**Review by Sacha0 — APPROVED** [review]
lgtm modulo the minor comment! :) (Caveat: I am not familiar the subtleties associated with this code. Someone familiar with those subtleties should have a look.)

File: base/irrationals.jl
Comment (giordano) [inline base/irrationals.jl]: I believe I already tried that and the wrong method was being picked up.  This is the same as `sinc(x::Complex{T}) where {T<:Integer}` in `special/trig.jl`.

Comment (Sacha0) [inline base/irrationals.jl]: Could you expand? :)
```julia
julia> Base.sinc(z::Complex{<:Irrational}) = "once more unto the breach"

julia> @which sinc(complex(pi, pi))
sinc(z::Complex{#s1} where #s1<:Irrational) in Main at REPL[1]:1

julia> sinc(complex(pi, pi))
"once more unto the breach"
```

Comment (musm) [inline base/irrationals.jl]: what sacha suggest is just sugar for what you have. if it is dispatching to the wrong method then it is a bug somewhere

Comment (giordano) [inline base/irrationals.jl]: Oh, didn't notice the `<:` before `Irrational`.  I tried `Complex{Irrational}`.  I'm going to adjust it.

Comment (giordano) [inline base/irrationals.jl]: Updated, thanks!

File: base/irrationals.jl
Comment (StefanKarpinski) [inline base/irrationals.jl]: Perhaps these should be defined statically and go through BigFloat for precision? Although I'm not sure MPFR exposes all of these functions for BigFloats.

Comment (giordano) [inline base/irrationals.jl]: Yes, there are methods for `sinpi` and `cospi` in MPFR.  You suggest something like this?
```julia
for c in (e, pi, catalan, golden)
    @eval begin
        Base.sinpi(x::typeof($c)) = $(Float64(sinpi(big(c))))
        Base.cospi(x::typeof($c)) = $(Float64(cospi(big(c))))
        Base.sinc(x::typeof($c)) = $(Float64(sinc(big(c))))
        Base.cosc(x::typeof($c)) = $(Float64(cosc(big(c))))
        Base.sinc(z::typeof(Complex($c, $c))) = $(Complex{Float64}(sinc(big(Complex(c, c)))))
        Base.cosc(z::typeof(Complex($c, $c))) = $(Complex{Float64}(cosc(big(Complex(c, c)))))
    end
end
```
I also thought about this, but the same can apply to all other functions and I don't think we're going to define a method for all functions in base library with all irrational numbers.  These definitions are anyway required to make custom irrational numbers work (otherwise users will need to define the methods themselves).  In any case, these methods should be more accurate than `function(pi * x)` (I compared with the bigfloat implementations), which is the goal of `*pi` functions.

Comment (StefanKarpinski) [inline base/irrationals.jl]: See, now this is a bit of a slippery slope. When do we stop implementing functions for Irrationals? Why these functions and not others?

Comment (giordano) [inline base/irrationals.jl]: Because other functions already works with irrationals! :-)  These are the only functions that I spotted that need to be explicitly defined for irrationals.  The problem is that methods of `sinpi` and `cospi` with argument `T<:Real` were probably implemented with `Integer` and `Rational` in mind, while `Irrational` are in a certain sense closer to `AbstractFloat`.

Comment (giordano) [inline base/irrationals.jl]: In this regard, an alternative is to change the signature of `sinpi(x::T) where T<:Real` to `sinpi(x::T) where T<:Union{Integer, Rationa}` and then define
```julia
sinpi(x::T) where T<:Real = sinpi(float(x))
```
so this wouldn't limited to irrationals, but would work with all other custom subtype of `Real`.  What do you think?

**Comment by giordano** [comment]
Following concerns by @StefanKarpinski, I reorganized the PR.  What I want to stress is that there is nothing special with `Irrational` type, it's just a subtype of `Real` that is sometimes overlooked in standard library.  By fixing functions for `Irrational` arguments, I'm (hopefully) making them work with any other custom subtype of `Real`.  I just want to make Julia more Julian ;-)

In the revised PR there is nothing specific to `Irrational`, only the tests are on `Irrational` type.  All tests in `test/math.jl` should pass.

File: base/special/trig.jl
Comment (tkelman) [inline base/special/trig.jl]: comment isn't accurate now

File: base/special/trig.jl
Comment (tkelman) [inline base/special/trig.jl]: also here

File: base/special/trig.jl
Comment (giordano) [inline base/special/trig.jl]: Fixed, also below.  Thanks!

**Comment by StefanKarpinski** [comment]
Thanks. This makes much more sense to me now!

File: base/irrationals.jl
Comment (giordano) [inline base/irrationals.jl]: @musm I'm a bit confused now: why
```julia
foo(x::Complex{Rational{<:Integer}}) = 0
```
isn't equivalent to
```julia
bar(x::Complex{Rational{T}}) where {T<:Integer} = 0
```
?

**Pull request merged**
